pkg/asset/installconfig/gcp/validation.go

```diff
@@ -469,31 +470,37 @@ func validateRegion(client API, ic *types.InstallConfig, fieldPath *field.Path)
 // ValidateCredentialMode checks whether the credential mode is
 // compatible with the authentication mode.
 func ValidateCredentialMode(client API, ic *types.InstallConfig) error {
-	creds := client.GetCredentials()
-
-	if creds.JSON == nil && ic.CredentialsMode != types.ManualCredentialsMode {
-		errMsg := "environmental authentication is only supported with Manual credentials mode"
-		return field.Forbidden(field.NewPath("credentialsMode"), errMsg)
-	}
-
-	return nil
+	allErrs := validateCredentialMode(client, ic)
+	return allErrs.ToAggregate()
 }
 
+// validateCredentialMode The presence of `authorized_user` in the credentials indicates that no service account
+// was used for authentication and requires Manual credential mode.
 func validateCredentialMode(client API, ic *types.InstallConfig) field.ErrorList {
 	allErrs := field.ErrorList{}
 
 	creds := client.GetCredentials()
-	if creds.JSON == nil {
-		if ic.CredentialsMode == "" {
-			logrus.Warn("Currently using GCP Environmental Authentication. Please set credentialsMode to manual, or provide a service account json file.")
-		} else {
-			if ic.CredentialsMode != "" && ic.CredentialsMode != types.ManualCredentialsMode {
-				errMsg := "environmental authentication is only supported with Manual credentials mode"
-				return append(allErrs, field.Forbidden(field.NewPath("credentialsMode"), errMsg))
-			}
+
+	if creds.JSON != nil {
+		var credsMap map[string]interface{}
+		err := json.Unmarshal(creds.JSON, &credsMap)
+		if err != nil {
+			return append(allErrs, field.Invalid(field.NewPath("credentials").Child("JSON"), creds.JSON, "failed to unmarshal JSON credentials"))
+		}
+
+		credsType, found := credsMap["type"]
+		if !found {
+			return append(allErrs, field.NotFound(field.NewPath("credentials").Child("JSON").Child("type"), "failed to find credentials type"))
 		}
-	}
 
+		if credsType.(string) == string(gcp.AuthorizedUserMode) && ic.CredentialsMode != types.ManualCredentialsMode {
+			errMsg := "environmental authentication is only supported with Manual credentials mode"
+			return append(allErrs, field.Forbidden(field.NewPath("credentialsMode"), errMsg))
+		}
+	} else if creds.JSON == nil && ic.CredentialsMode != types.ManualCredentialsMode {
+		errMsg := "environmental authentication is only supported with Manual credentials mode"
```

Comment:
How about "Manual credentials mode needs to be enabled to use environmental authentication"? We want to have different error messages for different error conditions 1. to help the user get to the correct configuration 2. for better debugging.

Reply:
Thanks for making the change even if you weren't totally convinced. :-)